### ISSUE TITLE
kubefed-controller-manager fails to start up after a clean helm installation on GKE 1.13.7-gke.8

### ISSUE THREAD

**Issue by queeno** [kubernetes-sigs/kubefed#1024]
**What happened**:
I have deployed a brand-new GKE cluster with two node pools on version `1.13.7-gke.8`.

I have followed the instructions of your [Helm chart deployment installation guide](https://github.com/kubernetes-sigs/kubefed/blob/master/charts/kubefed/README.md) in order to deploy kubefed `v0.1.0-rc3`.

The helm installation succeeds but when I run `kubectl get pods`, I notice that the `kubefed-controller-manager` pods die after 30s and are being continuously rebooted.
When catting the pod's logs, I can see the following output:

```
KubeFed controller-manager version: version.Info{Version:"v0.1.0-rc3", GitCommit:"d188d227fe3f78f33d74d9a40b3cb701c471cc7e", GitTreeState:"clean", BuildDate:"2019-06-25T00:46:45Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
W0701 19:38:40.161851       1 client_config.go:549] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0701 19:38:40.366123       1 controller-manager.go:206] Setting Options with KubeFedConfig "kube-federation-system/kubefed"
F0701 19:39:10.583889       1 controller-manager.go:320] Error updating KubeFedConfig "kube-federation-system/kubefed": Timeout: request did not complete within requested timeout 30s
```

A more in-depth investigation reveals that the error seems to happen on [this line](https://github.com/kubernetes-sigs/kubefed/blob/master/cmd/controller-manager/app/controller-manager.go#L318):

```go
err = client.Update(context.Background(), configResource)
if err != nil {
  klog.Fatalf("Error updating KubeFedConfig %q: %v", qualifiedName, err)
}
```

The `client.Update` command reports a timeout after 30s: `Timeout: request did not complete within requested timeout 30s`.

I have modified the source code to print out the `configResource` variable and the client configuration.

This is the `configResource`:

```
&{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:kubefed GenerateName: Namespace:kube-federation-system SelfLink:/apis/core.kubefed.k8s.io/v1beta1/namespaces/kube-federation-system/kubefedconfigs/kubefed UID:b16ee6e9-9c37-11e9-a35e-42010aac0003 ResourceVersion:27290 Generation:1 CreationTimestamp:2019-07-01 19:37:42 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Initializers:nil Finalizers:[] ClusterName:} Spec:{Scope:Cluster ControllerDuration:0xc00039a990 LeaderElect:0xc000567f00 FeatureGates:[{Name:PushReconciler Configuration:Enabled} {Name:SchedulerPreferences Configuration:Enabled} {Name:CrossClusterServiceDiscovery Configuration:Enabled} {Name:FederatedIngress Configuration:Enabled}] ClusterHealthCheck:0xc000567ec0 SyncController:0xc00000f418}}
```

And the client configuration:

```
&{Host:https://10.156.32.1:443 APIPath: ContentConfig:{AcceptContentTypes: ContentType: GroupVersion:<nil> NegotiatedSerializer:<nil>} Username: Password: BearerToken:eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJrdWJlLWZlZGVyYXRpb24tc3lzdGVtIiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9zZWNyZXQubmFtZSI6Imt1YmVmZWQtY29udHJvbGxlci10b2tlbi16aDR2OCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50Lm5hbWUiOiJrdWJlZmVkLWNvbnRyb2xsZXIiLCJrdWJlcm5ldGVzLmlvL3NlcnZpY2VhY2NvdW50L3NlcnZpY2UtYWNjb3VudC51aWQiOiJiMTIzOGI5Zi05YzM3LTExZTktYTM1ZS00MjAxMGFhYzAwMDMiLCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6a3ViZS1mZWRlcmF0aW9uLXN5c3RlbTprdWJlZmVkLWNvbnRyb2xsZXIifQ.hrlrQYhIX2iGQR8wr5Z7Dfa1CZl4u5rFEM0ECV-QJt2paj2YBSxdVP3BrRTdDBVGo7-PKrfvz5XosRGL8TSyYzHJrb391QjZgjPbK4JO1CIjKz4RbS2wY32t_8tskluBUOz3RKxrGv6r0HZpkNcPANQlQ-xuiSZHdVSFT8edxH7Guk23YzpSvwBrieOM-175jRHTaXwUsdW2Oy6E9eVUcvt70r_02LWAPWhflBWN7qXHwgbpu9VwH-Yh97wZU50bLVdx7swmXRXf7-GFazHbnWnTj00LcY69j-dByGaYLMeAfl1hSBqL0eqMz95uBNv-uRx1SwAtqaGxyHAM4YzF2Q BearerTokenFile:/var/run/secrets/kubernetes.io/serviceaccount/token Impersonate:{UserName: Groups:[] Extra:map[]} AuthProvider:<nil> AuthConfigPersister:<nil> ExecProvider:<nil> TLSClientConfig:{Insecure:true ServerName: CertFile: KeyFile: CAFile: CertData:[] KeyData:[] CAData:[]} UserAgent: Transport:<nil> WrapTransport:<nil> QPS:0 Burst:0 RateLimiter:<nil> Timeout:0s Dial:<nil>}
```

I can have also spun up a debug pod in my cluster and successfully run the following curl:

```
curl -k https://10.156.32.1/apis/core.kubefed.k8s.io/v1beta1/namespaces/kube-federation-system/kubefedconfigs/kubefed -H "Authoriz
ation: Bearer $TOKEN"
{"apiVersion":"core.kubefed.k8s.io/v1beta1","kind":"KubeFedConfig","metadata":{"creationTimestamp":"2019-07-01T19:37:42Z","generation":1,"name":"kubefed","namespace":"kube-federation-system","resourceVersion":"27290","selfLink":"/apis/core.kubefed.k8s.io/v1beta1/namespaces/kube-federation-system/kubefedconfigs/kubefed","uid":"b16ee6e9-9c37-11e9-a35e-42010aac0003"},"spec":{"clusterHealthCheck":{"failureThreshold":3,"period":"10s","successThreshold":1,"timeout":"3s"},"controllerDuration":{"availableDelay":"20s","unavailableDelay":"60s"},"featureGates":[{"configuration":"Enabled","name":"PushReconciler"},{"configuration":"Enabled","name":"SchedulerPreferences"},{"configuration":"Enabled","name":"CrossClusterServiceDiscovery"},{"configuration":"Enabled","name":"FederatedIngress"}],"leaderElect":{"leaseDuration":"15s","renewDeadline":"10s","resourceLock":"configmaps","retryPeriod":"5s"},"scope":"Cluster","syncController":{"adoptResources":"Enabled"}}}
```

**What you expected to happen**:

The kubefed-controller-manager pods should have successfully started up.

**How to reproduce it (as minimally and precisely as possible)**:

Spin up a new GKE cluster on the latest master and nodes versions (I have a private cluster configured with vpc-native networking). Deploy kubefed v0.1.0-rc3 with the helm chart as described above.

**Anything else we need to know?**:

Probably not, but feel free to ask.
Might be related to #928 

**Environment**:
- Kubernetes version (use `kubectl version`)
```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-20T04:49:16Z", GoVersion:"go1.12.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.7-gke.8", GitCommit:"7d3d6f113e933ed1b44b78dff4baf649258415e5", GitTreeState:"clean", BuildDate:"2019-06-19T16:37:16Z", GoVersion:"go1.11.5b4", Compiler:"gc", Platform:"linux/amd64"}
```
- KubeFed version: 0.1.0-rc3 (latest)
- Scope of installation (namespaced or cluster) cluster
- Others

/kind bug

**Comment by marun** [comment]
Thank you for filing! This issue is definitely unrelated to #928. The problem here is related to webhook misconfiguration that appears specific to GKE (or at least Kube 1.13 - we only test 1.14 in CI) rather than a broken image.  

The following slack conversation documents the failure to update KubeFedConfig as being caused by the validation webhook not being reachable even though the deployment running it and the service used to access it appear healthy:

https://kubernetes.slack.com/archives/C09R1PJR3/p1561534537123100

**Comment by queeno** [comment]
Hi @marun 

Thank you for your reply!

It took me a while to troubleshoot this one. First of all, I realised that I could increase the logging by setting an insanely high `-v=100` or so in the args of the pods. That provides an amazing insight in the HTTP requests being exchanged between the pods.

I have also realised that the timeout error that I am seeing when the controller manager starts up is the result of the admission control webhook trying to validate the creation of a new `kubefedconfig` objects. This request times out.

I have attempted to perform a `kubefedconfig` object creation using `kubectl` and I have observed the same timeout error. Had I known before that the timeout was connected to the admission webhook, it would have saved me a lot of time digging in the code.

In GKE, the Kubernetes control plane is deployed in a private account and there is a peering connection between the private master network and my GCP account. A firewall rule gets deployed automatically upon cluster creation allowing ports 443 and 10250 between the master network and the node pools.

I noticed that some traffic to the `kubefed-admission-webhook`'s Pod-specific IP address on port 8443 were being denied. This is really curious because I would expect the admission control webhook only to interact with the `kubefed-admission-webhook` service and never with the single pods.

As soon as I add port 8443 to the firewall rule allowing traffic between the master and the nodes, kubefed works. The admission control webhook changes its behaviour and now it fires its requests to the `kubefed-admission-webhook` service on port 443.

I have reproduced this behaviour in a clean GKE 1.13 environment with dummy resources and raised a bug with kubernetes/kubernetes#79739

I don't think this issue kubefed or even perhaps GKE specific. I look forward to new developments on the kubernetes issue.
